File: BackEnd/src/data/IMappers/IUserMapper.ts
/**
 * @file IUserMapper.ts
 * @description Interfaz del mapper de usuarios.
 * Define el contrato de transformación entre UserDocument (MongoDB) y User (entidad de dominio).
 * @module Auth
 * @layer Data
 */

import { User } from '../../domain/entities/User.js';
import { UserDocument } from '../datasources/mongodb/models/UserModel.js';

/**
 * Contrato del mapper de usuarios.
 * La implementación concreta vive en data/mappers/user_mapper.ts.
 */
export interface IUserMapper {
  /**
   * Convierte un documento MongoDB en una entidad de dominio User.
   *
   * @param doc — Documento tal como se almacena en MongoDB.
   * @returns Entidad User del dominio.
   */
  toEntity(doc: UserDocument): User;

  /**
   * Convierte una entidad User en un documento listo para insertar/actualizar en MongoDB.
   *
   * @param user — Entidad de dominio.
   * @returns Documento MongoDB (sin _id si es inserción nueva).
   */
  toDocument(user: User): Omit<UserDocument, '_id'>;
}

File: FrontEnd/lib/data/i_mappers/i_user_mapper.dart
/// @file i_user_mapper.dart
/// @description Interfaz del mapper de usuario. Define el contrato de conversión
/// entre UserModel (capa de datos) y User (entidad de dominio).
/// Los repositorios dependen de esta interfaz, no de la implementación concreta.
/// @module Core
/// @layer Data
library;

import '../../domain/entities/user.dart';
import '../models/user_model.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// I USER MAPPER
// ═══════════════════════════════════════════════════════════════════════════════

/// Contrato de conversión bidireccional UserModel ↔ User.
///
/// [implements] registrar implementación en container.dart.
abstract interface class IUserMapper {
  /// Convierte un [UserModel] (serialización de API) a una entidad [User] de dominio.
  User toEntity(UserModel model);

  /// Convierte una entidad [User] de dominio a [UserModel] para serialización.
  UserModel toModel(User entity);
}
